File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */
function buildTree(inorder: number[], postorder: number[]): TreeNode | null {
    const inMap = new Map<number, number>();
    
    for (let i = 0; i < inorder.length; i++) {
        inMap.set(inorder[i], i);
    }
    
    function buildTreeHelper(
        inStart: number,
        inEnd: number,
        postStart: number,
        postEnd: number
    ): TreeNode | null {
        if (inStart > inEnd) {
            return null;
        }
        
        const root = new TreeNode(postorder[postEnd]);
        const rootIndex = inMap.get(root.val)!;
        const leftSubtreeSize = rootIndex - inStart;
        
        root.left = buildTreeHelper(
            inStart,
            rootIndex - 1,
            postStart,
            postStart + leftSubtreeSize - 1
        );
        
        root.right = buildTreeHelper(
            rootIndex + 1,
            inEnd,
            postStart + leftSubtreeSize,
            postEnd - 1
        );
        
        return root;
    }
    
    return buildTreeHelper(0, inorder.length - 1, 0, postorder.length - 1);
}

File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.rs
use std::rc::Rc;
use std::cell::RefCell;
use std::collections::HashMap;

impl Solution {
    fn build_tree_helper(
        inorder: &[i32],
        in_start: i32,
        in_end: i32,
        postorder: &[i32],
        post_start: i32,
        post_end: i32,
        in_map: &HashMap<i32, i32>
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if in_start > in_end {
            return None;
        }
        
        let mut root = TreeNode::new(postorder[post_end as usize]);
        let root_index = in_map[&postorder[post_end as usize]];
        let left_subtree_size = root_index - in_start;
        
        root.left = Self::build_tree_helper(
            inorder,
            in_start,
            root_index - 1,
            postorder,
            post_start,
            post_start + left_subtree_size - 1,
            in_map
        );
        
        root.right = Self::build_tree_helper(
            inorder,
            root_index + 1,
            in_end,
            postorder,
            post_start + left_subtree_size,
            post_end - 1,
            in_map
        );
        
        Some(Rc::new(RefCell::new(root)))
    }
    
    pub fn build_tree(inorder: Vec<i32>, postorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut in_map = HashMap::new();
        for (i, &val) in inorder.iter().enumerate() {
            in_map.insert(val, i as i32);
        }
        
        if inorder.is_empty() {
            return None;
        }
        
        Self::build_tree_helper(
            &inorder,
            0,
            (inorder.len() - 1) as i32,
            &postorder,
            0,
            (postorder.len() - 1) as i32,
            &in_map
        )
    }
}